commandapi-nms/commandapi-1.19-common/src/main/java/dev/jorel/commandapi/nms/NMS_1_19_Common.java

```diff
@@ -577,6 +536,111 @@ public final ParticleData<?> getParticle(CommandContext<CommandSourceStack> cmdC
+			destination = getVibrationParticleOptionDestinationAsBlockPositionSource(positionSource, level);
+		} else if (options.getDestination() instanceof EntityPositionSource positionSource) {
+			destination = getVibrationParticleOptionDestinationAsEntityPositionSource(positionSource, level);
+		} else {
+			CommandAPI.getLogger().warning("Unknown vibration destination " + options.getDestination());
+			return new ParticleData<Void>(particle, null);
+		}
+		return new ParticleData<Vibration>(particle, new Vibration(from, destination, options.getArrivalInTicks()));
+	}
+	
+	/**
+	 * {@link #getParticleDataAsVibrationParticleOption(CommandContext, Level, Particle, VibrationParticleOption)
+	 * Parent Method}
+	 * @param positionSource {@link BlockPositionSource} the block position source to find the designation of.
+	 * @param level {@link Level} The level in which to find the destination
+	 * @return {@link Destination} the destination of the provided source in the given level.
+	 */
+	private @NotNull Destination getVibrationParticleOptionDestinationAsBlockPositionSource(@NotNull BlockPositionSource positionSource, Level level) {
+		Vec3 to = positionSource.getPosition(level).get();
+		return new BlockDestination(new Location(level.getWorld(), to.x(), to.y(), to.z()));
+	}
+
+	/**
+	 * {@link #getParticleDataAsVibrationParticleOption(CommandContext, Level, Particle, VibrationParticleOption)
+	 * Parent Method}
+	 * @param positionSource {@link EntityPositionSource} the entity position source to find the designation of.
+	 * @param level {@link Level} The level in which to find the destination
+	 * @return {@link Destination} the destination of the provided source in the given level.
+	 */
+
+	private @NotNull Destination getVibrationParticleOptionDestinationAsEntityPositionSource(@NotNull EntityPositionSource positionSource, Level level) {
+		positionSource.getPosition(level); // Populate Optional sourceEntity
+		Either<Entity, Either<UUID, Integer>> entity = (Either<Entity, Either<UUID, Integer>>) EntityPositionSource_sourceEntity
+			.get(positionSource);
+		if (entity.left().isPresent()) {
+			return new EntityDestination(entity.left().get().getBukkitEntity());
+		}
+		Either<UUID, Integer> id = entity.right().get();
+		if (id.left().isPresent()) {
+			return new EntityDestination(Bukkit.getEntity(id.left().get()));
+		}
+
+		// Do an entity lookup by numerical ID. There's no "helper function"
+		// like Bukkit.getEntity() to do this for us, so we just have to do it
+		// manually
+		Entity foundEntity = null;
+
+		// Asset Minecraft Server is not null to stop null pointer exception.
+		assert MINECRAFT_SERVER != null;
+		for (ServerLevel world : MINECRAFT_SERVER.getAllLevels()) {
+			Entity entityById = world.getEntity(id.right().get());
+			if (entityById != null) {
+				foundEntity = entityById;
+				break;
+			}
+		}
+		// Asset found entity is not null to stop null pointer exception.
+		assert foundEntity != null;
```

Comment:
We can assume `foundEntity` is not null. This is actually checked in Minecraft's game packets when it triggers a vibration particle. In the rare case that it doesn't catch this, an exception will be thrown and caught and shown to users (that something went wrong when trying to input that particle argument), which means that they probably entered the wrong entity ID or something which is perfectly fine.

Please remove this assertion.